### PR TITLE
feat: show approval editor for permit2 messages

### PR DESCRIPTION
**Pull request by schmanu**
## What it solves
Permit2 signature are not highlighted in our UI.

## How this PR fixes it
Shows the readonly approval editor for `PermitSingle` and `PermitBatch` Permit2 messages in SignOnchain and Offchain modals

## How to test it
- Request a signature for a PermitSingle or PermitBatch message (I extended the [example drainer app](https://github.com/5afe/eip-1271-dapp/tree/drain-wallet-example-dapp))
- Observe the new Approval Editor

## Screenshots
Screenshot 2024-01-12 at 16 48 10

Screenshot 2024-01-12 at 18 03 37

## Checklist
* [x] I've tested the branch on mobile 📱
* [x] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/components/tx/ApprovalEditor/index.tsx [hunk 1]
```diff
@@ -29,17 +30,26 @@ const Title = () => {
...
 }
 
-export const ApprovalEditor = ({ safeTransaction }: { safeTransaction: SafeTransaction | undefined }) => {
+export const ApprovalEditor = ({
+  safeTransaction,
+  safeMessage,
+}: {
+  safeTransaction?: SafeTransaction
+  safeMessage?: EIP712TypedData
+}) => {
```
Comment (usame-algan) [inline src/components/tx/ApprovalEditor/index.tsx:39]: With `safeMessage` added here we have to extend a lot of the conditions and it propagates down through `useApprovalInfos` to the `ApprovalModule` and adds more complexity.

Afaiu if there is a `safeMessage` we are always in readOnly so we could split this into two components, one for transactions one for safe messages and add a new method `ApprovalModule.scanMessage` to handle this case and check inside `useApprovalInfos` in one place if `safeMessage ? ApprovalModule.scanMessage : ApprovalModule.scanTransaction`. Wdyt?

Comment (schmanu) [reply src/components/tx/ApprovalEditor/index.tsx:39]: I think adding a scanMessage function helps readability of the `ApprovalModule` but I do not think we need to create a separate component as most of the components logic is shared / we use the same type for Permits as for Approvals.

I will refactor it tomorrow a bit more :)

**Review by usame-algan — APPROVED** [review]
I would suggest to extend the tests for `ApprovalEditor` to cover the new conditions

- Should return null if there is no safeMessage
- Should be read-only if there is a safeMessage

File: src/services/security/modules/ApprovalModule/index.ts [hunk 2]
```diff
@@ -52,11 +60,59 @@ export class ApprovalModule implements SecurityModule<ApprovalModuleRequest, App
     return []
   }
 
-  scanTransaction(request: ApprovalModuleRequest): SecurityResponse<ApprovalModuleResponse> {
-    const { safeTransaction } = request
-    const safeTxData = safeTransaction.data.data
+  private static scanPermitDetails(details: PermitDetails): Pick<Approval, 'amount' | 'tokenAddress'> {
```
Comment (usame-algan) [inline src/services/security/modules/ApprovalModule/index.ts:63]: Would it make more sense to name this `getPermitDetails`? Its not really scanning something.

File: src/components/tx/ApprovalEditor/hooks/useApprovalInfos.test.ts [hunk 3]
```diff
@@ -95,6 +97,219 @@ describe('useApprovalInfos', () => {
...
+          },
+        ],
+      },
+    }
+
+    const { result } = renderHook(() => useApprovalInfos({ safeMessage: mockMessage }))
+
+    const expectedApprovals: ApprovalInfo[] = [
+      {
+        amount: getTokenInfo.UNLIMITED_PERMIT2_AMOUNT.toBigInt(),
```
Comment (compojoom) [inline src/components/tx/ApprovalEditor/hooks/useApprovalInfos.test.ts:291]: everywhere else i see Bignumber used, but here we do toBigInt(). Why?

**Comment by francovenica** [comment]
So I checked with the dummy app and the approval editor shows up just fine in this PR. For comparison I've checked in prod and there it doesn't show up at all,  so with this I'd be inclined to pass the ticket.

Before passing it a question:
Currently in this scenario, the approval editor is read only and fixed in unlimited. Is this because of the dummy app? does every app send a specific amount that it needs to be allowed to use? or at least leave the field editable for the user to change?

This PR:
image

Prod:
image

File: src/components/tx/ApprovalEditor/hooks/useApprovalInfos.test.ts (hunk 3 above)
Comment (schmanu) [reply src/components/tx/ApprovalEditor/hooks/useApprovalInfos.test.ts]: Good point. I think I had too many context switched to ethers v6. Replaced it with BigNumber!

**Comment by schmanu** [comment]
> Currently in this scenario, the approval editor is read only and fixed in unlimited. Is this because of the dummy app? 

@francovenica yes this is the case. In the test dApp I fixed the amounts to Unlimited. Lower amounts work too though. I once edited them in the dummy App.

> does every app send a specific amount that it needs to be allowed to use? or at least leave the field editable for the user to change?
It would not work to edit these values as this could not be communicated to the dApp. The dApp will only receive a signature in return. The dApp is responsible for the amount that is being sent. It could also send a lower amount than unlimited.

**Pull request merged**